index.d.ts

```diff
@@ -97,13 +97,13 @@ declare module 'robot3' {
 
   /* General Types */
 
-  export type ContextFunction<T> = (context: T, event: unknown) => T
+  export type ContextFunction<T> = (initialContext: T) => any
 
-  export type GuardFunction<T> = (context: T, event: unknown) => boolean
+  export type GuardFunction<C, E> = (context: C, event: E) => boolean
 
-  export type ActionFunction<T> = (context: T, event: unknown) => boolean
+  export type ActionFunction<C, E> = (context: C, event: E) => void
```

Comment:
The `ActionFunction` needs to return true or a truthy value because the implementation is using the `reduce` function type to return context. 

```js
export let action = fn => reduce((ctx, ev) => !!~fn(ctx, ev) && ctx);
```

```suggestion
  export type ActionFunction<C, E> = (context: C, event: E) => boolean
```

Reply:
OK. I wasn't clear on what the `~` is for here but see that one effect is that a function that returns void will return true. In this case, the input function itself does not return a truthy value. We can't constrain the return type of `ActionFunction` to be a boolean because most actions won't return true or false. What do you think about the return type being `unknown`?

Reply:
Just tested with `unknown` and this resolves the issue I mentioned about the context type defaulting to `boolean | void` if not explicitly set.

Reply:
`unknown` makes sense. Thanks for testing that out.